Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```diff
@@ -0,0 +1,274 @@
+        enum PollLoopAction {
+            /// Ask `librdkakfa` to receive new message acknowledgements through.
+            case poll
+            /// Suspend the poll loop.
+            case suspendPollLoop
+            /// Shutdown the poll loop.
+            case shutdownPollLoop
+        }
+
+        /// The events that can be triggered by the ``KafkaBackPressurePollingSystem``.
+        enum Event {
+            /// Produce more elements.
+            case produceMore
+            /// Suspend the poll loop until the continuation is resumed.
+            case suspendLoop(CheckedContinuation<Void, Never>)
+            /// Our `Task` was cancelled why we were waiting for the loop to be unsuspended.
+            case loopSuspensionCancelled
+            /// Request to stop producing elements. When this reaches the poll loop it will escalate
+            /// to a full suspension.
+            case stopProducing
+            /// Close the entire ``KafkaBackPressurePollingSystem`` with its poll loop.
+            case shutdown
+        }
+
+        /// The possible states of the state machine.
+        enum State {
+            /// Initial state.
+            case initial
+            /// The system up and producing acknowledgement messages.
+            case producing
+            /// We were asked to stop producing messages. However we still wait for the
+            /// poll loop to be suspended.
+            case suspended
+            /// The pool loop is currently suspended and we are waiting for an invocation
+            /// of `produceMore()` to continue producing messages.
+            case loopSuspended(CheckedContinuation<Void, Never>)
+            /// The system is shut down.
+            case shutDown
+        }
+
+        /// The current state of the state machine.
+        let state = NIOLockedValueBox(State.initial)
+
+        /// Determines the next action to be taken in the poll loop based on the current state.
+        ///
+        /// - Returns: The next action for the poll loop.
+        func nextPollLoopAction() -> PollLoopAction {
+            self.state.withLockedValue {
+                switch $0 {
+                case .initial, .producing:
+                    return .poll
+                case .suspended:
+                    // We were asked to stop producing,
+                    // but the poll loop is still running.
+                    // Trigger the poll loop to suspend.
+                    return .suspendPollLoop
+                case .loopSuspended:
+                    fatalError("Illegal state: cannot invoke \(#function) when poll loop is suspended")
+                case .shutDown:
+                    return .shutdownPollLoop
+                }
+            }
+        }
+
+        /// Triggers an event in the state machine, causing a state transition.
+        ///
+        /// - Parameter event: The event to be triggered.
+        func eventTriggered(_ event: Event) {
```

Comment:
Event triggered is way too broad. Please introduce separate methods for all the various events you model here e.g. `pollSuspended`, `pollCancelled` etc.